## Agentic Workflow with RAG
Author: **Peeyush Sharma**; Feedback: **PSharma3@gmail.com**

This notebooks combines Agentic AI workflow with Retrieval Augmented Generation (RAG). We extract financial returns of a public company through a series of prompts in a single workflow to the point where the data can be seamlessly imported in an Excel file. Here is a sample automated extraction:

Balance Sheet:
| Timestamp | Metric | Value |
|-----------|---------|--------|
| March 31st, 2025 | Total assets | $1,950,311 million |
| March 31st, 2025 | Total loans | $913,842 million |
| March 31st, 2025 | Total deposits | $1,361,728 million |
| March 31st, 2025 | Total equity | $182,906 million |
| December 31st, 2024 | Total assets | $1,929,845 million |
| December 31st, 2024 | Total loans | $912,745 million |
| December 31st, 2024 | Total deposits | $1,371,804 million |
| December 31st, 2024 | Total equity | $181,066 million |

To skip directly to the final results, compare the extractions at the bottom of this notebook with major financial returns captured in the following 10-Q submission: https://www.wellsfargo.com/assets/pdf/about/investor-relations/sec-filings/2025/first-quarter-10q.pdf. Otherwise, read along for detailed analysis.

Agentic AI is usually interpreted in two different ways:
1. An autonomous set of agents that proactively monitor a condition and can take action without explicit human intervention.
2. A workflow of parallel, sequential, or iterative steps where each step is governed by a prompt and an LLM agent.

This notebook demonstrate the 2nd use-case above in conjunction with RAG. To achieve that, a very detailed sequence of prompts has been used to sequentially retrieve, cleanse, and format data attributes. This specific sample covers quarterly filing of a public firm but can be extended for any data extraction use-case. The first prompt is run against "similarity" embeddings for commonly reported financial terms. All subsequent prompts are then called on the output of the previous prompt iteration resulting in a nicely formatted table-like structure that can easily be imported into Excel as delimited CSV.

In [ ]:
import os
from uuid import uuid4
from dotenv import load_dotenv

import chromadb
import torch
from langchain.chains import RetrievalQA
from langchain.document_loaders import PyPDFLoader
from langchain.text_splitter import CharacterTextSplitter
from langchain_chroma import Chroma
from langchain_huggingface import HuggingFaceEmbeddings
from typing import List, Dict, Callable
from langchain_ollama import OllamaLLM
from langchain_core.prompts import ChatPromptTemplate
from langchain_anthropic import AnthropicLLM
from langchain_anthropic import ChatAnthropic

from util import llm_call, extract_xml


In [ ]:
# Basic setup/config items
DOC_DIR = "../documents"
DOC_TYPE = "10-Q"
EMBEDDING_MODEL_NAME = "all-MiniLM-L6-v2"
# LLM_MODEL_NAME = "llama3.2" # Local LLM with Ollama
LLM_MODEL_NAME = "claude-3-5-sonnet-20241022" # Cloud LLM with Anthropic Claude
CHUNK_SIZE = 4096
CHUNK_OVERLAP = 256
PURGE_VECTOR_DB = True # Purging everytime for demo env

In [ ]:
# Embedding DB definition and vector stores
device = torch.device("cuda:0" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "cpu")
embeddings = HuggingFaceEmbeddings(model_name=EMBEDDING_MODEL_NAME)
collection_name = "rag_collection"
chroma_client = chromadb.PersistentClient()

try:
    if PURGE_VECTOR_DB: # Keep option for hard reset
        chroma_client.delete_collection(name=collection_name)
    chroma_client.get_or_create_collection(collection_name)
except ValueError:
    chroma_client.delete_collection(collection_name)

collection = chroma_client.get_or_create_collection(collection_name)

vector_store = Chroma(
    client=chroma_client,
    collection_name=collection_name,
    embedding_function=embeddings
)

In [ ]:
# Function to split the document into chunks and embed them
def insert_document_to_vector_db(doc_path):
    loader = PyPDFLoader(doc_path)
    documents = loader.load()
    text_splitter = CharacterTextSplitter(chunk_size=CHUNK_SIZE, chunk_overlap=CHUNK_OVERLAP, separator="\n")
    docs = text_splitter.split_documents(documents=documents)
    uuids = [str(uuid4()) for _ in range(len(docs))]
    # print(len(uuids))
    ids = vector_store.add_documents(documents=docs, ids=uuids)
    return ids

In [19]:
# Embedding target file
dir_path = os.path.join(DOC_DIR, DOC_TYPE)
vector_indices = []
# Add optionality to cover only the
# suggested files within the folder
file_list = ['wfc-first-quarter-10q.pdf',]
             #'wfc-first-quarter-10q-loan-section.pdf']
if os.path.exists(dir_path) and os.path.isdir(dir_path):
    temp_indices = [insert_document_to_vector_db(os.path.join(dir_path, doc))
                    for doc in os.listdir(dir_path)
                    if doc in file_list]
    vector_indices.extend(temp_indices)
# vector_indices

In [27]:
def workflow(input: str, prompts: List[str]) -> str:
    """Chain multiple LLM calls sequentially, passing results between steps."""
    documentation = input
    # llm = OllamaLLM(model=LLM_MODEL_NAME)
    llm = ChatAnthropic(model=LLM_MODEL_NAME)
    for i, prompt in enumerate(prompts, 1):
        print(f"\nStep {i}:")
        if i == 1:
            retriever = vector_store.as_retriever(search_type="similarity",
                                                  search_kwargs={'k': 30})
            docs = retriever.invoke(prompt)
            # Format the docs as a string to pass to call to model
            documentation = "\n".join([doc.page_content for doc in docs])
        # Calls made to local Lllama 3.2
        prompt = ChatPromptTemplate.from_template(template=prompt)
        chain = prompt | llm
        documentation = chain.invoke({"documentation": documentation, "question": prompt})

        # Calls for Anthropic - commented out with poor results
        # llm_input = f"""{prompt}.\n Here is the documentation: \n{documentation}\n"""
        # documentation = llm_call(llm_input)
        print(f"Result AFTER step {i}: {documentation.content}")
    return documentation.content

In [28]:
# The actual magic happens here. Each step progressively
# transforms raw text into a well formatted table like structure
data_processing_steps = [
    """You are a financial analyst. You are given excerpts from a single 10-Q statement
     for a public company: {documentation}. You are to identify all major financial data attributes from
    cash flow, balance sheet, and income statements from this 10-Q document and
    return that data to the user. Keep the originally reported format
    in your output as much as possible.
    Example format:
    (in millions, except share data)
    March 31, 2025
    December 31,2024
    Assets
    Cash and due from banks
    $
    22,066
    $
    23,372
    Deposits with banks
    403,837
    445,945
    """,

    """Segregate references by balance sheet, income statement, and cash flows
    in the following documentation: {documentation}. Do not remove any data.
    Example format:

    Balance Sheet:
        March 31, 2025
        December 31,2024
        Assets
        Cash and due from banks
        $
        22,066
        $
        23,372
        Deposits with banks
        403,837
        445,945
        ...
        ...
    Cash Flow:
         March 31, 2025
        December 31,2024
        Assets
        Cash and due from banks
        $
        22,066
        $
        23,372
        Deposits with banks
        403,837
        445,945
        ...
        ...
    Income Statement:
         March 31, 2025
        December 31,2024
        Assets
        Cash and due from banks
        $
        22,066
        $
        23,372
        Deposits with banks
        403,837
        445,945
        ...
        ...
    """,

    """Extract the numerical values, fiancial term they represent, and associated
     currency ($) and unit (millions, thousands) from the given documentation:
    {documentation}. Do not remove any data points. Keep data segregated by
    balance sheet, cash flows, and income statement. The dots (...) below mean
    that there are more data points under this category.
    Example format:
    Balance Sheet:
    Quarter ending March 31st, 2025: Outstanding loans: (473,571) million
    Quarter ending December 31st, 2024: Credit Losses: (12, 634) million
    ...
    ...
    Cash flows:
    Quarter ending March 31st, 2025: Outstanding loans: (473,571) million
    Quarter ending December 31st, 2024: Credit Losses: (12, 634) million
    ...
    ...
    Income statement:
    Quarter ending March 31st, 2025: Outstanding loans: (473,571) million
    Quarter ending December 31st, 2024: Credit Losses: (12, 634) million
    ...
    ...
    """,

    """Format each data point as 'quarter: metric: currency value units' on a new line
    in the following documentation: {documentation}.
    Keep one number per line. Do not remove any data points.
    Keep data segregated by
    balance sheet, cash flows, and income statement.
    Example format:
    March 31st, 2025: Cash and due from banks: $22,066 million
    March 31st, 2025: Deposit with banks: $403,837 thousand
    ...
    ...
    """,

    """Format the data as a markdown table with columns
    in the following documentation: {documentation}.
    Do not remove any data points. Keep data segregated by
    balance sheet, cash flows, and income statement.
    If you have any other categories beyond these three, then
    merge that category's data into balance sheet data.
    Example format:

    Balance Sheet:
    Timestamp | Metric | Value
    March 31st, 2025 | Customer Satisfaction | 92%
    December 31st, 2024 | Loans and leases | $934,311 thousand

    Cash Flow:
    Timestamp | Metric | Value
    March 31st, 2025 | Deposits with banks | (40,3817) thousand
    December 31st, 2024 | Loans originated, net of principal collected | (-1,721) million

    Timestamp | Metric | Value
    December 31st, 2024 | Deposits with banks | (44,5995) thousand
    March 31st, 2025 | Cash outflows for investing activities | (-2,135) million

    """
]

# Make the actual call. The input text is empty
# as the source is read from embedding DB.
# Keeping the text option here for demo purposes
formatted_result = workflow('', data_processing_steps)



Step 1:
Result AFTER step 1: Based on the 10-Q document, here are the key financial data attributes:

Consolidated Income Statement Data (in millions, except per share amounts):
Quarter ended March 31,
2025 2024
Net interest income $ 11,495 $ 12,227
Noninterest income 8,654 8,636
Total revenue 20,149 20,863
Provision for credit losses 932 938
Noninterest expense 13,891 14,338
Net income 4,894 4,619
Net income applicable to common stock 4,616 4,313

Per Share Information:
Earnings per common share $ 1.41 $ 1.21
Diluted earnings per common share 1.39 1.20
Average common shares outstanding 3,280.4 3,560.1
Diluted average common shares outstanding 3,321.6 3,600.1

Balance Sheet Data (period-end, in millions):
March 31, 2025 Dec 31, 2024
Total assets $ 1,950,311 $ 1,929,845
Total loans 913,842 912,745
Total deposits 1,361,728 1,371,804
Total equity 182,906 181,066

Cash Flow Data (in millions):
Quarter ended March 31,
2025 2024
Net cash used by operating activities $ (11,037) $ (12,125)
Ne

Balance Sheet:
| Timestamp | Metric | Value |
|-----------|---------|--------|
| March 31st, 2025 | Total assets | $1,950,311 million |
| March 31st, 2025 | Total loans | $913,842 million |
| March 31st, 2025 | Total deposits | $1,361,728 million |
| March 31st, 2025 | Total equity | $182,906 million |
| December 31st, 2024 | Total assets | $1,929,845 million |
| December 31st, 2024 | Total loans | $912,745 million |
| December 31st, 2024 | Total deposits | $1,371,804 million |
| December 31st, 2024 | Total equity | $181,066 million |


Income Statement:
| Timestamp | Metric | Value |
|-----------|---------|--------|
| March 31st, 2025 | Net interest income | $11,495 million |
| March 31st, 2025 | Noninterest income | $8,654 million |
| March 31st, 2025 | Total revenue | $20,149 million |
| March 31st, 2025 | Provision for credit losses | $932 million |
| March 31st, 2025 | Noninterest expense | $13,891 million |
| March 31st, 2025 | Net income | $4,894 million |
| March 31st, 2025 | Net income applicable to common stock | $4,616 million |
| March 31st, 2025 | Earnings per common share | $1.41 |
| March 31st, 2025 | Diluted earnings per share | $1.39 |
| March 31st, 2025 | Average common shares | 3,280.4 million |
| March 31st, 2025 | Diluted average common shares | 3,321.6 million |
| March 31st, 2024 | Net interest income | $12,227 million |
| March 31st, 2024 | Noninterest income | $8,636 million |
| March 31st, 2024 | Total revenue | $20,863 million |
| March 31st, 2024 | Provision for credit losses | $938 million |
| March 31st, 2024 | Noninterest expense | $14,338 million |
| March 31st, 2024 | Net income | $4,619 million |
| March 31st, 2024 | Net income applicable to common stock | $4,313 million |
| March 31st, 2024 | Earnings per common share | $1.21 |
| March 31st, 2024 | Diluted earnings per share | $1.20 |
| March 31st, 2024 | Average common shares | 3,560.1 million |
| March 31st, 2024 | Diluted average common shares | 3,600.1 million |

| Timestamp | Metric | Value |
|-----------|---------|--------|
| March 31st, 2025 | Net cash used by operating activities | $(11,037) million |
| March 31st, 2025 | Net cash provided (used) by investing activities | $(27,508) million |
| March 31st, 2025 | Net cash provided by financing activities | $12,821 million |
| March 31st, 2025 | Net change in cash, cash equivalents and restricted cash | $(25,724) million |
| March 31st, 2024 | Net cash used by operating activities | $(12,125) million |
| March 31st, 2024 | Net cash provided (used) by investing activities | $21,095 million |
| March 31st, 2024 | Net cash provided by financing activities | $23,574 million |
| March 31st, 2024 | Net change in cash, cash equivalents and restricted cash | $32,544 million |